### PR TITLE
Matrix multiplication for CUDA batched and non-batched int32/int64 tensors.

### LINKED ISSUE #44428: Cuda support for matrix multiplication on Long Tensor
## 🚀 Feature
I would love cuda support for matrix multiplication on Long Tensor
Currently:
```
import torch

x = torch.LongTensor([[1]])
x = x.cuda()
x + x  # works
x * x  # works
x @ x  # FAILS
```

## Motivation & Pitch

I'm working on privacy preserving machine learning which uses a lot LongTensors (like the CrypTen team cc @jspisak ), and having matmul accelerated with cuda (like it is for normal matmul on floats) would be game changing.

Is there any plan to support this, or any work around?

Thanks a lot for all the efforts.

### PR DESCRIPTION
**Pull request by v0dro**
Fixes https://github.com/pytorch/pytorch/issues/44428.

Adds direct support for matrix multiplication for batched and non-batched tensors using 32 and 64 bit integer data types on CUDA.

### CODE REVIEW

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 1]
```diff
@@ -168,25 +169,55 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
...
-      transpose_mat1 ? mat1_->is_conj() ? 'c' : 't' : 'n',
-      transpose_mat2 ? mat2_->is_conj() ? 'c' : 't' : 'n',
-      m, n, k,
-      alpha_val,
-      mat1_ptr, mat1_ld,
-      mat2_ptr, mat2_ld,
-      beta_val,
-      result_ptr, result_ld
+  // TODO: is it possible to combine the dispatch code?
+  if (isIntegralType(scalar_type, false)) {
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/Blas.cpp:173]: `isIntegralType` evaluates `(t == ScalarType::Byte || t == ScalarType::Char || t == ScalarType::Int ||
       t == ScalarType::Long || t == ScalarType::Short)` but you have implementation of `batched_integer_gemm` only for ` ScalarType::Int` and `ScalarType::Long`.

Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/Blas.cpp:173]: You could place the new code into a separate function and have in this function:
```cpp
if (is_int32_or_int64) {
call_integer_specialization();
return result;
}
```
And the diff would be cleaner and no need to make the existing function more complicated to quickly glance through.

File: aten/src/ATen/native/cuda/Blas.cu [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#include <ATen/cuda/CUDABlas.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <c10/macros/Macros.h>
+
+namespace at {
+namespace cuda {
+namespace blas {
+
+  template <typename scalar_t>
+  __global__ void gemm_kernel(int64_t M, int64_t N, int64_t K, scalar_t alpha, const scalar_t* a, int64_t lda,
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/Blas.cu:10]: Since this is a new file you have an opportunity to use a more descriptive name, something like "IntegerMatmul.cu" instead of vague "Blas". Maybe "IntegerBlas" if the plan is to implement integer support also for other functions from BLAS API.

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 1 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/Blas.cpp:173]: Works.

File: aten/src/ATen/native/cuda/Blas.cu (hunk 2 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/Blas.cu]: Since there's no plan for other integer BLAS operations I'll confine it to IntegerMatmul.cu for now.

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 3]
```diff
@@ -168,25 +169,50 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
             c10::nullopt /* pin_memory */));
   }
 
-  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!result_->is_conj());
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/Blas.cpp:171]: Why did you remove this internal assert?

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 4]
```diff
@@ -168,25 +169,50 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
...
-    at::cuda::blas::gemm<scalar_t>(
-      transpose_mat1 ? mat1_->is_conj() ? 'c' : 't' : 'n',
-      transpose_mat2 ? mat2_->is_conj() ? 'c' : 't' : 'n',
-      m, n, k,
-      alpha_val,
-      mat1_ptr, mat1_ld,
-      mat2_ptr, mat2_ld,
-      beta_val,
-      result_ptr, result_ld
+  // TODO: is it possible to combine the dispatch code?
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/Blas.cpp:172]: I think so, as long as you call `integer_gemm` simply `gemm`. Since integer tensors do not have a float bit, the handling of `'c'` should be fine.

My only complaint would be that the function would live in the `blas::` namespace, even though it's not a `cuBLAS` function. This is a bit annoying, but it's not terribly bad.

File: aten/src/ATen/native/cuda/IntegerMatmul.cu [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+#include <ATen/native/cuda/Loops.cuh>
+#include <c10/macros/Macros.h>
+
+namespace at {
+namespace cuda {
+namespace blas {
+
+  template <typename scalar_t, bool transA=false, bool transB=false>
+  __global__ void gemm_kernel(int64_t M, int64_t N, int64_t K, scalar_t alpha,
+                                                  const scalar_t* a, int64_t lda, const scalar_t * b, int64_t ldb,
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/IntegerMatmul.cu:13]: Making `a` and `b` `const scalar_t* __restrict__` should hopefully improve the performance, as `a` and  `b` would be saved into the L1 cache. This should be good as there quite a big number of repeated accesses in the kernel.

**Review by lezcano — COMMENTED** [review]
Left a number of comments

**Comment by ngimel** [comment]
Can you please post benchmarks for sizes ranging from say 512 to a few K, and compare with fp32 perf?

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 4 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: It does work:
```c++
#include <iostream>

template<typename scalar_t>
scalar_t f1(scalar_t x) {
  std::cout << "calling f1" << std::endl;
  return x;
}

template<typename scalar_t>
scalar_t f2(scalar_t x) {
  std::cout << "calling f2" << std::endl;
  return x;
}

template<typename scalar_t>
struct F {
  static decltype(&f1<scalar_t>) f(scalar_t x) {
    if (x < 3) return f1<scalar_t>;
    return f2<scalar_t>;
  }
};

int main(int argn, char* args[]) {
  F<int>::f(2)(3);
  F<int>::f(5)(4);
}

```
Outputs 
```
calling f1
calling f2

```

Comment (Lezcano) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: Oh, I see what you mean now. Yes, you are right, this works. You could do something like:
```c++
const bool is_int = scalar_type == ScalarType::Long || scalar_type == ScalarType::Int;

auto gemm = is_int ? &at::cuda::blas::integer_gemm<scalar_t> 
                                : &at::cuda::blas::gemm<scalar_t>;
```

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 6]
```diff
@@ -261,24 +287,58 @@ Tensor& baddbmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor&
...
+  if (self.scalar_type() == ScalarType::Long || self.scalar_type() == ScalarType::Int) {
+    AT_DISPATCH_INDEX_TYPES(self.scalar_type(), "baddmm_cuda_int", [&] {
+      using scalar_t = index_t;
+
+      scalar_t alpha_val = alpha.to<scalar_t>();
+      scalar_t beta_val = beta.to<scalar_t>();
+
+      auto batch1_stride = batch1_->strides()[0];
+      auto batch2_stride = batch2_->strides()[0];
+      auto result_stride = result_->strides()[0];
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/Blas.cpp:299]: From these strides it looks like 3D tensors are assumed. What if I a feed an n-dim stride with n > 3? Looks like a reshape is missing somewhere, or am I wrong?

Comment (Lezcano) [reply aten/src/ATen/native/cuda/Blas.cpp:299]: `bmm` accepts just 3D tensors. The reshape is done in `matmul`.

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 7]
```diff
@@ -261,24 +287,58 @@ Tensor& baddbmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor&
...
+
+      auto batch1_data = batch1_->data_ptr<scalar_t>();
+      auto batch2_data = batch2_->data_ptr<scalar_t>();
+      auto result_data = result_->data_ptr<scalar_t>();
+
+      for (int64_t batch = 0; batch < num_batches; ++batch) {
+        scalar_t * batch1_working_ptr = &batch1_data[batch * batch1_stride];
+        scalar_t * batch2_working_ptr = &batch2_data[batch * batch2_stride];
+        scalar_t * result_working_ptr = &result_data[batch * result_stride];
+          at::cuda::blas::integer_gemm<scalar_t>(
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/Blas.cpp:309]: This loop will work if the batch dimensions are contiguous. Do you check for each input `x` if `(x.contiguous() or x.t().contiguous())` evaluates to `True`?

File: test/test_linalg.py [hunk 8]
```diff
@@ -6072,6 +6072,51 @@ def test_matmul_45724(self, device):
...
+
+                self.assertEqual(int_prod, fp_prod)
+
+    @onlyCUDA
+    @dtypes(torch.int32, torch.int64)
+    def test_matmul_batched_integer_tensor(self, device, dtype):
+        for batch_size in [1, 100, 500]:
+            m = 20
+            n = 30
+            k = 10
```
Comment (nikitaved) [inline test/test_linalg.py:6109]: Could you test for multi-dim batch dimension which is also non-contiguous?

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 4 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: How do you change the `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2` macro and other macros like it so that they can dispatch to integer types too? From what I can see you still need to have separate FP/complex+FP16 macros along with an extra call to `AT_DISPATCH_INDEX_TYPES` if you want to dispatch to `int32` and `int64`.

The above solution would reduce duplication in terms of code within the macros but I'll still need to call two separate macros.

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 7 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/Blas.cpp:309]: I've added a check in the code and will add a test in the test files too.

File: aten/src/ATen/native/cuda/IntegerMatmul.cu [hunk 9]
```diff
@@ -0,0 +1,77 @@
+#include <ATen/cuda/CUDABlas.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/cuda/detail/KernelUtils.h>
+#include <ATen/native/cuda/Loops.cuh>
+#include <c10/macros/Macros.h>
+
+namespace at {
+namespace cuda {
+namespace blas {
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply aten/src/ATen/native/cuda/IntegerMatmul.cu:9]: In my opinion, I think it'd be better in a different namespace, and just leave that of blas for calls to blas functions as to make clear what is it that we are calling. The functions already live in different files and have different names, so it's clear when are we calling to an implementation of our own or to an external implementation

File: aten/src/ATen/native/cuda/IntegerMatmul.cu (hunk 5 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/IntegerMatmul.cu:13]: Done.

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 6 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/Blas.cpp:299]: Agreed. The code above this function checks whether the output is a result of a transpose operation or not and will then change the stride dimension of the result. This approach is working the `bgemm` call to cuBLAS below.

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 4 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: I bet there must dispatch keys for almost all types, I guess. You can check indexing operators as they work for all dtypes.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: I believe that dispatch at floating/complex and two more that you specify should do the job.

Comment (v0dro) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: Its already using B16 as the two more data types.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: there must AND3 macro, right?
```
pytorch-cuda-dev) nik@qgpu3:~/git/Quansight/pytorch$ git grep AND3                                                                                                                                               
/usr/bin/pager: /home/nik/.conda/envs/pytorch-cuda-dev/lib/libtinfo.so.6: no version information available (required by /usr/bin/pager)                                                                           
aten/src/ATen/Dispatch.h:#define AT_DISPATCH_ALL_TYPES_AND3(                                         \                                                                                                            
aten/src/ATen/Dispatch.h:#define AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                             \                                                                                                            
aten/src/ATen/ScalarOps.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                                             
aten/src/ATen/Utils.cpp:AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)                                                                                                                                 
aten/src/ATen/native/BinaryOps.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                                      
aten/src/ATen/native/Copy.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kHalf, kBool, kBFloat16, self.scalar_type(), "copy_", [&] {                                                                                
aten/src/ATen/native/ReduceOps.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBool, kBFloat16, kHalf, iter.input_dtype(), "equal_cpu", [&] {
aten/src/ATen/native/Scalar.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                                         
aten/src/ATen/native/TensorAdvancedIndexing.cpp:    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16,                                                              
aten/src/ATen/native/TensorAdvancedIndexing.cpp:      AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16,                                                            
aten/src/ATen/native/TensorAdvancedIndexing.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                         
aten/src/ATen/native/TensorAdvancedIndexing.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                         
aten/src/ATen/native/TensorAdvancedIndexing.cpp:  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(                                                                                                                         
aten/src/ATen/native/TensorFactories.cpp:AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, DEFINE_CAST_OP)      
...
```

Comment (nikitaved) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: Hmm, indeed, there is no such macro. What about writing a new version with AND3?

**Comment by v0dro** [comment]
@ngimel so I just benchmarked my integer-mult implementation vs. casting to float/double and using cuBLAS and that seems to be much cheaper in terms of time. Here's the results:

```
[--------------------------------------------------------- matmul --------------------------------------------------------]
                            |  torch.int32 product  |  torch.int64 product  |  torch.float product  |  torch.double product
1 threads: ----------------------------------------------------------------------------------------------------------------
      (2, 10, 10, 20)       |            11         |            11         |            32         |            22        
      (2, 40, 80, 50)       |            21         |            30         |            24         |            86        
      (2, 100, 200, 100)    |           247         |           304         |            36         |           453        
      (4, 10, 10, 20)       |            14         |            20         |            24         |            23        
      (4, 40, 80, 50)       |            49         |            62         |            24         |           200        
      (4, 100, 200, 100)    |           493         |           605         |            70         |           880        
      (8, 10, 10, 20)       |            22         |            31         |            24         |            22        
      (8, 40, 80, 50)       |            80         |           100         |            24         |           255        
      (8, 100, 200, 100)    |          1000         |          1000         |           100         |          2000        
      (16, 10, 10, 20)      |            40         |            61         |            24         |            22        
      (16, 40, 80, 50)      |           200         |           300         |            49         |           522        
      (16, 100, 200, 100)   |          2000         |          2447         |           243         |          3264        
      (32, 10, 10, 20)      |            80         |           100         |            22         |            35        
      (32, 40, 80, 50)      |           385         |           620         |            91         |           930        
      (32, 100, 200, 100)   |          3865         |          5255         |           450         |          6410        
      (64, 10, 10, 20)      |           160         |           250         |            24         |            61        
      (64, 40, 80, 50)      |           730         |          1000         |           170         |          1800        
      (64, 100, 200, 100)   |          7700         |         10000         |           900         |         10000        
      (128, 10, 10, 20)     |           305         |           620         |            42         |           130        
      (128, 40, 80, 50)     |          1460         |          2505         |           330         |          3628        
      (128, 100, 200, 100)  |         15000         |         21760         |          1790         |         25480        
      (256, 10, 10, 20)     |           700         |          1500         |           100         |           266        
      (256, 40, 80, 50)     |          2967         |          4770         |           642         |          7200        
      (256, 100, 200, 100)  |         30740         |         39770         |          3513         |         52000   
```

The benchmarking script is like so:
<details>

``` python

import torch
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import itertools

batches = [2, 4, 8, 16, 32, 64, 128, 256]
types = [torch.int32, torch.int64, torch.float, torch.double]
result_shapes = [(10, 10, 20), (40, 80, 50), (100, 200, 100)]
results = []

for t in [torch.int32, torch.int64]:
    for batch, shape in itertools.product(batches, result_shapes):
        m, n, k = shape
        mat1 = torch.randint(0, 100, (batch, m, k), device='cuda', dtype=t)
        mat2 = torch.randint(400, 1000, (batch, k, n), device='cuda', dtype=t)
        label = "matmul"

        timer = Timer(stmt="mat1 @ mat2", num_threads=1, label=label, sub_label=f"({batch}, {m}, {n}, {k})",
                      description=f"{t} product", globals=globals())

        results.append(timer.blocked_autorange())

for t in [torch.int32, torch.int64]:
    for batch, shape in itertools.product(batches, result_shapes):
        m, n, k = shape
        mat1 = torch.randint(0, 100, (batch, m, k), device='cuda', dtype=t)
        mat2 = torch.randint(400, 1000, (batch, k, n), device='cuda', dtype=t)
        label = "matmul"

        if t == torch.int32:
            timer = Timer(stmt="mat1.to(torch.float) @ mat2.to(torch.float)", num_threads=1,
                          label=label, sub_label=f"({batch}, {m}, {n}, {k})",
                          description="torch.float product", globals=globals())
        elif t == torch.int64:
            timer = Timer(stmt="mat1.to(torch.double) @ mat2.to(torch.double)", num_threads=1,
                          label=label, sub_label=f"({batch}, {m}, {n}, {k})",
                          description="torch.double product", globals=globals())

        results.append(timer.blocked_autorange())

compare = Compare(results)
compare.trim_significant_figures()
compare.colorize()
compare.print()
```
</details>

Apart from precision, I don't see any real benefit of using integer-specific matmul routines. 

@Lezcano suggests using `BlockReduceSum` from `aten/src/ATen/native/cuda/block_reduce.cu` but I'm not sure if it can make up for the massive difference between cuBLAS.

**Comment by v0dro** [comment]
INT64 is competitive with FP64 in some cases but the rest are pretty much lagging behind severely.

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 4 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: We'll need AND4 (int32 + int64 + Bfloat16 + FP16). Is that justified for not having a single if-else in only one part of the code?

**Comment by v0dro** [comment]
Another test with the offloading to CPU and then getting the tensor back on the GPU:

Code:
<details>

``` python
import torch
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import itertools

batches = [2, 4, 8, 16, 32, 64, 128]
types = [torch.int32, torch.int64, torch.float, torch.double]
result_shapes = [(10, 10, 20), (40, 80, 50)]
results = []

for t in [torch.int32, torch.int64]:
    for batch, shape in itertools.product(batches, result_shapes):
        m, n, k = shape
        mat1 = torch.randint(0, 100, (batch, m, k), device='cuda', dtype=t)
        mat2 = torch.randint(400, 1000, (batch, k, n), device='cuda', dtype=t)
        label = "matmul"

        timer = Timer(stmt="mat1 @ mat2", num_threads=1, label=label, sub_label=f"({batch}, {m}, {n}, {k})",
                      description=f"{t} product", globals=globals())

        results.append(timer.blocked_autorange())

for t in [torch.int32, torch.int64]:
    for batch, shape in itertools.product(batches, result_shapes):
        m, n, k = shape
        mat1 = torch.randint(0, 100, (batch, m, k), device='cuda', dtype=t)
        mat2 = torch.randint(400, 1000, (batch, k, n), device='cuda', dtype=t)
        label = "matmul"

        timer = Timer(stmt="(mat1.to('cpu') @ mat2.to('cpu')).to('cuda')",
                      num_threads=1, label=label, sub_label=f"({batch}, {m}, {n}, {k})",
                      description=f"{t} cuda->cpu->cuda", globals=globals())

        results.append(timer.blocked_autorange())

for t in [torch.int32, torch.int64]:
    for batch, shape in itertools.product(batches, result_shapes):
        m, n, k = shape
        mat1 = torch.randint(0, 100, (batch, m, k), device='cuda', dtype=t)
        mat2 = torch.randint(400, 1000, (batch, k, n), device='cuda', dtype=t)
        label = "matmul"

        if t == torch.int32:
            timer = Timer(stmt="mat1.to(torch.float) @ mat2.to(torch.float)", num_threads=1,
                          label=label, sub_label=f"({batch}, {m}, {n}, {k})",
                          description="torch.float product", globals=globals())
        elif t == torch.int64:
            timer = Timer(stmt="mat1.to(torch.double) @ mat2.to(torch.double)", num_threads=1,
                          label=label, sub_label=f"({batch}, {m}, {n}, {k})",
                          description="torch.double product", globals=globals())

        results.append(timer.blocked_autorange())

compare = Compare(results)
compare.trim_significant_figures()
compare.colorize()
compare.print()
```
</details>

Benchmarks:

```
[--------------------------------------------------------------------------------------- matmul ---------------------------------------------------------------------------------------]
                         |  torch.int32 product  |  torch.int64 product  |  torch.int32 cuda->cpu->cuda  |  torch.int64 cuda->cpu->cuda  |  torch.float product  |  torch.double product
1 threads: -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      (2, 10, 10, 20)    |            11         |            11         |                43             |                48             |           33          |            22        
      (2, 40, 80, 50)    |            22         |            32         |               200             |               433             |           23          |            86        
      (4, 10, 10, 20)    |            14         |            20         |                50             |                56             |           22          |            24        
      (4, 40, 80, 50)    |            44         |            70         |               354             |               810             |           23          |           200        
      (8, 10, 10, 20)    |            24         |            34         |                60             |                76             |           22          |            22        
      (8, 40, 80, 50)    |            90         |           200         |               667             |              1600             |           23          |           258        
      (16, 10, 10, 20)   |            41         |            62         |                88             |               100             |           22          |            22        
      (16, 40, 80, 50)   |           190         |           298         |              1300             |              3110             |           49          |           506        
      (32, 10, 10, 20)   |            80         |           100         |               100             |               190             |           22          |            35        
      (32, 40, 80, 50)   |           370         |           588         |              2547             |              6180             |           91          |          1000        
      (64, 10, 10, 20)   |           157         |           246         |               240             |               335             |           24          |            61        
      (64, 40, 80, 50)   |           700         |          1200         |              5098             |             12000             |          200          |          2000        
      (128, 10, 10, 20)  |           331         |           589         |               433             |               626             |           43          |           100        
      (128, 40, 80, 50)  |          2000         |          2382         |              9970             |             24270             |          333          |          3685      
```

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 4 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: Well, we can indeed keep it in an if-else part but it is still possible to compress the logic in a single function, right?
Something like
```
template <typename scalar_t>
void logic(...)

if (...) {
  AT_DISPATCH_FLOATING_AND_COMPLEX(..., logic<scalar_t>());
}
else {
 AT_DISPATCH_INDEX_TYPE(..., logic<scalar_t>());
}
```

Comment (nikitaved) [reply aten/src/ATen/native/cuda/Blas.cpp:172]: I actually wonder why AT_DISPATCH is not implemented with variadics...

**Comment by mruberry** [comment]
Hey v0dro! Thanks for the benchmarks. I was discussing this with @ngimel and based on the available benchmarks we think we should just encourage people to stick with converting their tensors to double before the matmul. @ngimel points out that on non-consumer cards double performance will likely be significantly better, too, further widening the performance gap observed above. 

Maybe in the future CUDA math libraries will offer integer matmuls or we'll use new technologies like Triton to generate integer CUDA matmul kernels that are competitive with double CUDA matmul kernels.

**Comment by Lezcano** [comment]
@mruberry @ngimel Casting to double is not an option in the use case of https://github.com/pytorch/pytorch/issues/44428, as they need to work with integers when dealing with privacy / cryptography related applications. Exact computations are completely necessary in these fields.

As such, the first benchmark is not really representative, as casting to `double` does not solve  It is the second one that matters, in which it is shown that this implementation improves on the current best method of doing `(a.cpu() @ b.cpu()).cuda()` when `a`, `b` are `int32` or `int64`. In my opinion, having an in-house integer matmul for CUDA, given that it improves on the casting to CPU and back to CUDA method, I think that could be very useful.

**Comment by v0dro** [comment]
@mruberry so what's the consensus on this PR? Is it still necessary to implement exact integer matmul specifically for the GPU?

**Comment by Lezcano** [comment]
It does not seem like we'll land this. It looks like we won't be supporting integer matmul for quite some time given the lack of activity in the relevant issue, and the fact that we don't have an efficient and correct implementation on CPU or CUDA. As such, I'm closing this PR.

If we ever support this, I think it'll be through NVIDIA's cutlass.
